pkg/util/external_storage.go

```diff
@@ -212,3 +229,52 @@ func IsNotExistInExtStorage(err error) bool {
 	}
 	return false
 }
+
+// RemoveFilesIf removes files from external storage if the path matches the predicate.
+func RemoveFilesIf(
+	ctx context.Context,
+	extStorage storage.ExternalStorage,
+	pred func(path string) bool,
+	opt *storage.WalkOption,
+) error {
+	var toRemoveFiles []string
+	err := extStorage.WalkDir(ctx, opt, func(path string, _ int64) error {
+		path = strings.TrimPrefix(path, "/")
+		if pred(path) {
+			toRemoveFiles = append(toRemoveFiles, path)
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.ErrExternalStorageAPI.Wrap(err).GenWithStackByArgs("RemoveTemporaryFiles")
+	}
+
+	log.Debug("Removing files", zap.Any("toRemoveFiles", toRemoveFiles))
+
+	for _, path := range toRemoveFiles {
+		if err := extStorage.DeleteFile(ctx, path); err != nil {
+			return errors.ErrExternalStorageAPI.Wrap(err)
+		}
+	}
+	return DeleteFilesInExtStorage(ctx, extStorage, toRemoveFiles)
+}
+
+// DeleteFilesInExtStorage deletes files in external storage concurrently.
+func DeleteFilesInExtStorage(
+	ctx context.Context, extStorage storage.ExternalStorage, toRemoveFiles []string,
+) error {
+	eg, eCtx := errgroup.WithContext(ctx)
+	for _, file := range toRemoveFiles {
```

Comment:
Will the `toRemoveFiles` be very large? How many files will TiCDC generate in a flush interval?

Reply:
Yes, there could be a lot of files here and we need to limit concurrency here.